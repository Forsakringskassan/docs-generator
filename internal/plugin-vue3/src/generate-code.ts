import {
    type BindingMetadata,
    compileScript,
    compileStyle,
    compileTemplate,
    parse,
} from "vue/compiler-sfc";

export interface ExampleOptions {
    /* original filename */
    filename: string;
    /* example slug */
    slug: string;
    /* example fingerprint (hash of the sourcecode) */
    fingerprint: string;
    /** example sourcecode */
    code: string;
    /** absolute path to setup function */
    setupPath: string;
}

export interface ExampleResult {
    /* HTML markup to inject into the document */
    markup: string;
    /* JS sourcecode to compile */
    sourcecode: string;
    /* asset filename */
    output: string;
}

export function generateCode(options: ExampleOptions): ExampleResult {
    const { slug, fingerprint, code, filename, setupPath } = options;
    const selector = `#${slug}`;
    const asset = `${slug}-${fingerprint}.js`;
    const { descriptor, errors } = parse(code, { filename });
    const scopeId = `data-v-${fingerprint}`;

    if (errors.length) {
        throw new Error(`Errors occured when trying to parse ${filename}.`);
    }

    const hasScoped = descriptor.styles.some((e) => e.scoped);
    const styleContent = descriptor.styles
        .map((stylePart) => {
            const compiledStyle = compileStyle({
                filename: descriptor.filename,
                source: stylePart.content,
                isProd: false,
                id: scopeId,
                scoped: stylePart.scoped,
                trim: true,
            });
            return compiledStyle.code;
        })
        .join("\n");

    let sourcecode = `import { setup } from "${setupPath}";\n`;

    let bindings: BindingMetadata | undefined = undefined;
    if (descriptor.script || descriptor.scriptSetup) {
        const script = compileScript(descriptor, {
            id: scopeId,
            isProd: false,
            sourceMap: false,
            inlineTemplate: false,
            genDefaultAs: "exampleComponent",
            templateOptions: {
                id: scopeId,
                filename,
                source: descriptor.template?.content,
                scoped: hasScoped,
                slotted: descriptor.slotted,
                compilerOptions: {
                    comments: false,
                    whitespace: "condense",
                    scopeId: hasScoped ? scopeId : undefined,
                    mode: "module",
                },
            },
        });
        bindings = script.bindings;
        sourcecode += script.content;
        sourcecode += "\n\n";
    } else {
        /* edge case: handle when the vue component does not have <script> but only a <template> */
        sourcecode += `const exampleComponent = {};\n\n`;
    }

    if (descriptor.template) {
        const template = compileTemplate({
            id: scopeId,
            filename,
            source: descriptor.template.content,
            scoped: hasScoped,
            slotted: descriptor.slotted,
            preprocessLang: descriptor.template.lang,
            compilerOptions: {
                bindingMetadata: bindings,
                comments: false,
                whitespace: "condense",
                scopeId: hasScoped ? scopeId : undefined,
                mode: "module",
            },
        });
        sourcecode += `${template.code}\n\n`;
        sourcecode += `exampleComponent.render = render;\n`;
    }

    if (hasScoped) {
        sourcecode += `exampleComponent.__scopeId = "${scopeId}";\n`;
    }

    sourcecode += `\nsetup({\n  rootComponent: exampleComponent,\n  selector: "${selector}"\n});\n`;

    const markup = /* HTML */ `
        <div id="${slug}"></div>
        <!-- [html-validate-disable-next element-permitted-content -- technical debt, should be inserted into <head>] -->
        <style>
            ${styleContent}
        </style>
        <!-- [html-validate-disable-next require-sri -- technical debt, should generate integrity but we don't know the hash of the compiled file yet] -->
        <script defer src="./${asset}"></script>
    `;

    return {
        markup,
        sourcecode,
        output: asset,
    };
}
