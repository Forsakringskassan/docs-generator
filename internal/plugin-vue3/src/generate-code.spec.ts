import fs from "node:fs";
import path from "node:path";
import { type ExampleOptions, generateCode } from "./generate-code";

const fixturePath = path.join(__dirname, "__fixtures__");

function readFixture(filename: string): ExampleOptions {
    const slug = filename
        .replace(".vue", "")
        .replace(/(^|.)([A-Z])/g, (_, p1: string, p2: string) => {
            return `${p1}${p1 ? "-" : ""}${p2.toLowerCase()}`;
        });
    const fingerprint = "123456";
    const setupPath = "/path/to/setup.ts";
    const fixture = fs.readFileSync(path.join(fixturePath, filename), "utf-8");
    return {
        filename,
        slug,
        fingerprint,
        setupPath,
        code: fixture,
    };
}

it("should generate markup", () => {
    expect.assertions(1);
    const { markup } = generateCode(readFixture("MockExample.vue"));
    expect(markup).toMatchSnapshot();
});

it("should generate compiled SFC", () => {
    expect.assertions(1);
    const { sourcecode } = generateCode(readFixture("MockExample.vue"));
    expect(sourcecode).toMatchSnapshot();
});

it("should handle missing <script>", () => {
    expect.assertions(1);
    const { sourcecode } = generateCode(readFixture("NoScript.vue"));
    expect(sourcecode).toMatchSnapshot();
});
