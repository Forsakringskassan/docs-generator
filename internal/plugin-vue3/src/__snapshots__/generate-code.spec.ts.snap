// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate compiled SFC 1`] = `
"import { setup } from "/path/to/setup.ts";

import { defineComponent } from "vue";

const exampleComponent = defineComponent({
    name: "MockExample",
    props: {
        name: {
            type: String,
            required: true,
        },
    },
});


import { toDisplayString as _toDisplayString, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue"

const _hoisted_1 = { class: "greeting" }

export function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock("p", _hoisted_1, "Hello " + _toDisplayString(_ctx.name) + "!", 1 /* TEXT */))
}

exampleComponent.render = render;

setup({
  rootComponent: exampleComponent,
  selector: "#example-d661fd"
});
"
`;

exports[`should generate markup 1`] = `
"
        <div id="example-a92dca"></div>
        <!-- [html-validate-disable-next element-permitted-content -- technical debt, should be inserted into <head>] -->
        <style>
            
.greeting {
    color: red;
}

        </style>
        <!-- [html-validate-disable-next require-sri -- technical debt, should generate integrity but we don't know the hash of the compiled file yet] -->
        <script type="module" src="./mock-example-a92dca.js"></script>
    "
`;

exports[`should handle missing <script> 1`] = `
"import { setup } from "/path/to/setup.ts";
const exampleComponent = {};

import { openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue"

const _hoisted_1 = { class: "greeting" }

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock("p", _hoisted_1, "Hello World!"))
}

exampleComponent.render = render;

setup({
  rootComponent: exampleComponent,
  selector: "#example-1877f3"
});
"
`;
